### ISSUE TITLE
Best way to test Api.js?

### ISSUE THREAD

**Issue by alexsoble** [studentinsights/studentinsights#1327]
I added the first test file for Api.js in #1326. Right now it's very minimal, it just tests whether the class decides correctly between create note/update note. 

I'd like to add more thorough tests, but I ran into a couple of questions that I'd love to talk about with other folks who are interested in JS testing choices! 

**Injecting dependencies.** I started experimenting with Jest mocks, but quickly found them hard to inject into the code. How do folks feel about refactoring so we can inject our HTTP-request-sending code as dependencies? 

This would be something like:

```
// current

  _updateNote(studentId, eventNoteParams) {
    return this._put('/students/' + studentId + '/event_notes/' + id + '.json', {
      event_note: {
        ...
        text: eventNoteParams.text,
      }
    });
  }

// refactor

  _updateNote(studentId, eventNoteParams, putFn) {
    return putFn('/students/' + studentId + '/event_notes/' + id + '.json', {
      event_note: {
        ...
        text: eventNoteParams.text,
      }
    });
  }
```

That way we could pass in `this._put` in the feature side code, and pass in a Jest mock in the test code. More flexible! What do folks think? 

**Private functions.** Some of the meatiest logic in Api.js is in `_createNote` and `_updateNote`. I know the leading underscore denotes a private function, and I know private functions generally aren't tested. So I'm not sure what to make of that! 

At first I thought that these functions had a leading underscore because they use jQuery instead of fetch as their API, and that was a "likely to change" choice that we wanted to hide away as an implementation detail. But `_createNote` and `_updateNote` don't call either jQuery or fetch, instead they pass data on to `_post` and `_put`. 

tl;dr I want to add tests for `_createNote` and `_updateNote` but I'm pausing because they're private functions, and maybe there's something I'm missing about the intent of the code! @kevinrobinson I believe you wrote these so let me know if you have anything to add. Clearly not a high-priority topic, just programming discussion.

**Comment by alexsoble** [comment]
The more refactoring I do, the more I see this pattern: for example, with functions that directly call `window.location.href = ` without making it injectable. Is this a style/pattern we want to encourage or enforce around the codebase, "Make all calls to external services and `window` injectable"?

**Comment by kevinrobinson** [comment]
@alexsoble Yeah, this is a great observation!  My concrete recommendation is: let's solve this as minimally as possible along the way to finish the ES6 module work and avoid trying to come up with a holistic solution across the whole app.

I noticed https://github.com/studentinsights/studentinsights/pull/1326/files#r158039410 on the pull request already merged, which will change the mechanics for how we test this as well.

re: injecting the dependency, this is a great strategy for testing, particularly with classes!  Another is to mock the call to `$.ajax` in the test, and call the public methods and see if they make the HTTP requests we expect.  That would be my suggestion for this particular case since the Api methods right now are written to be called statically (eg, there wasn't a `new Api` or if there is it's just thrown out each time).

So as a suggestion for an overall heuristic: For testing static methods or plain function, trying mock the dependency.  If it's an object or component, try updating the code to pass in the dependency so you can do the same in tests.

Here, one of the root contexts for this problem is that `Api` is written as stateless (meaning that any caller can just call these methods), which is mirroring and wrapping how `$.ajax` works.  Callers aren't instantiating the object and passing around an `api` instance everywhere.  If that were the case, then it'd make more sense to pass in the dependency like you say.  For updating how that works in the code, that's a larger and potentially good change but really dependent on the particular object and usage.  So for `window.location.href` we might look at making those calls through React Router instead and testing through that, whereas for `$.ajax` we ultimately need to mock somewhere before we get to an HTTP call.

Anyway, hope this is helpful thoughts and these are great things to be thinking about how to improve! 🌮 👍

**Comment by kevinrobinson** [comment]
oh and sorry, I read the Api module wrong - it was exporting a class previously.  So injecting would make more sense - but injecting the dependencies into the constructor (eg, `new Api({ ajax: $.ajax })`) rather than in the method call.

Here for the purpose of testing, updating the `Api` constructor to allow an optional `ajax` param for testing sounds like a great idea! (after un-Reacting it)

**Comment by alexsoble** [comment]
@kevinrobinson These are super helpful thoughts! Closing this issue now since as you point out, we should focus on the ES6 module-ifying work first and solve this as minimally as possible for the time being.
